pkg/validation/validation-rule.md

```diff
@@ -0,0 +1,58 @@
+### Id and Name:
+`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+
+The restriction is added to allow easy translation to K8s resource names, and also to have consistent rules for both name and id fields
+The validation will be done as part of schema validation, the rule will be introduced as a regex in schema definition, any objection of the rule in devfile will result in a failure.
+
+- limit to lowercase characters i.e.; no uppercase allowed
+- limit within 63 characters, except for endpoint’s name which is limited within 15 characters
+- no special characters allowed except dash(-)
+- start with an alphanumeric character
+- end with an alphanumeric character
+- must not be with all numbers (not included in schema validation, will have extra validation function)
+
+
+### Endpoints:
+- all the endpoint names are unique across components
+Since network is shared in the same pod, endpoint ports should be unique across components, two components cannot have the same target port but two endpoints in a single component can have the same target port. Only exception: container component with `dedicatedpod=true`
+
+### Commands:
+1. id must be unique
+2. composite command:
+    - Should not reference itself via a subcommand
+    - Should not indirectly reference itself via a subcommand which is a composite command
+    - Should reference a valid devfile command
+    - Subcommands should be valid as well
```

Comment:
I'm not sure this should be a required check--unless there's different constraints on subcommands checking each command should be sufficient. An error saying "composite command X is invalid: subcommand Y is invalid... command Y is invalid" is less clear than just "command Y is invalid"